### PR TITLE
Add shadow functionality to publishing process

### PR DESCRIPTION
**Pull request by alexander-schranz**
This implements the shadowing behaviour like discussed with @chirimoya. The data of the `draft` page stays as it is and when publishing a `Shadow` page the content from the ShadowLocale will be copied to the published page.

## TODO Functional Tests

 - [x] DE (Published: Not Shadow, Unpublished: Shadow to EN) EN Publish should change nothing
 - [x] DE (Published: Nothing, Unpublished: Shadow to EN) EN Publish should change nothing
 - [x] DE (Published: Shadow to EN, Unpublished: Not Shadow) EN Publish should update Live Shadow
 - [x] DE (Published: Shadow to EN, Unpublished: Shadow to EN) DE Unpublish EN Publish should publish nothing for DE
 
## TODO Unit Tests

 - [x] Update PublishTransitionSubscriberTest for new cases
 - [x] Update ContentCopierTests for new Params
 - [x] Update UnublishTransitionSubscriberTest for new cases
 - [x] Update ShadowDataMapperTests for new cases

### CODE REVIEW

File: Content/Infrastructure/Doctrine/MetadataLoader.php [hunk 1]
```diff
@@ -60,6 +60,18 @@ public function loadClassMetadata(LoadClassMetadataEventArgs $event): void
             $this->addIndex($metadata, 'idx_stage', ['stage']);
         }
 
+        if ($reflection->implementsInterface(ShadowInterface::class)) {
```
Comment (alexander-schranz) [inline Content/Infrastructure/Doctrine/MetadataLoader.php:63]: Just moved this to change the ordering of the columns as I think it make sense to have this columns visible first in the database table view before seo, excerpt, ...

File: Content/Infrastructure/Sulu/Admin/ContentViewBuilderFactory.php [hunk 2]
```diff
@@ -236,6 +240,12 @@ public function createViews(
                 $settingsToolbarActions,
                 $dimensionContentClass
             );
+
+            foreach ($views as $view) {
+                if ($view instanceof PreviewFormViewBuilderInterface) {
+                    $view->setPreviewCondition('shadowOn != true');
+                }
+            }
```
Comment (alexander-schranz) [inline Content/Infrastructure/Sulu/Admin/ContentViewBuilderFactory.php:248]: Quickfix as the preview currently can not handle shadow feature.

File: Content/Application/ContentCopier/ContentCopierInterface.php [hunk 3]
```diff
@@ -26,14 +26,18 @@ interface ContentCopierInterface
      * @param mixed[] $sourceDimensionAttributes
      * @param ContentRichEntityInterface<T> $targetContentRichEntity
      * @param mixed[] $targetDimensionAttributes
+     * @param mixed[] $data This data is merged with the data of the source content before set on the target content
+     * @param string[] $ignoredAttributes This attributes stayed untouched
```
Comment (Prokyonn) [inline Content/Application/ContentCopier/ContentCopierInterface.php]:
```suggestion
     * @param string[] $ignoredAttributes These attributes stay untouched
```

File: Content/Application/ContentCopier/ContentCopierInterface.php [hunk 4]
```diff
@@ -42,13 +46,17 @@ public function copy(
      * @param DimensionContentCollectionInterface<T> $dimensionContentCollection
      * @param ContentRichEntityInterface<T> $targetContentRichEntity
      * @param mixed[] $targetDimensionAttributes
+     * @param mixed[] $data This data is merged with the data of the source content before set on the target content
+     * @param string[] $ignoredAttributes This attributes stayed untouched
```
Comment (Prokyonn) [inline Content/Application/ContentCopier/ContentCopierInterface.php]:
```suggestion
     * @param string[] $ignoredAttributes These attributes stay untouched
```

File: Content/Application/ContentCopier/ContentCopierInterface.php [hunk 5]
```diff
@@ -57,12 +65,16 @@ public function copyFromDimensionContentCollection(
      * @param T $dimensionContent
      * @param ContentRichEntityInterface<T> $targetContentRichEntity
      * @param mixed[] $targetDimensionAttributes
+     * @param mixed[] $data This data is merged with the data of the source content before set on the target content
+     * @param string[] $ignoredAttributes This attributes stayed untouched
```
Comment (Prokyonn) [inline Content/Application/ContentCopier/ContentCopierInterface.php]:
```suggestion
     * @param string[] $ignoredAttributes These attributes stay untouched
```

File: Content/Application/ContentDataMapper/DataMapper/TemplateDataMapper.php [hunk 6]
```diff
@@ -75,44 +77,47 @@ public function map(
...
             $name = $property->getName();
 
             // Float are converted to ints in php array as key so we need convert it to string
             if (\is_float($name)) {
                 $name = (string) $name;
             }
 
-            if (\array_key_exists($name, $data)) {
+            $value = $property->isLocalized() ? $defaultLocalizedData[$name] ?? null : $defaultLocalizedData[$name] ?? null;
+            if (\array_key_exists($name, $data)) { // values not explicitly given need to stay untouched for e.g. for shadow pages urls
```
Comment (Prokyonn) [inline Content/Application/ContentDataMapper/DataMapper/TemplateDataMapper.php:120]:
```suggestion
            if (\array_key_exists($name, $data)) { // values not explicitly given need to stay untouched for e.g. shadow pages urls
```

File: Tests/Functional/Integration/ExampleControllerAvailableAndShadowLocalesTest.php [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+
+/**
+ * This test is responsibility to test the update of available and shadow locales based on different
+ * publishing processes.
+ *
+ * The integration test should have no impact on the coverage so we set it to coversNothing.
+ *
+ * @coversNothing
+ */
+class ExampleControllerAvailableAndShadowLocalesTest extends SuluTestCase
```
Comment (alexander-schranz) [inline Tests/Functional/Integration/ExampleControllerAvailableAndShadowLocalesTest.php:29]: As it is a very individual test case with different tests depending on each other I thought it most make sense to have it in an own class

File: Tests/Functional/Integration/ExampleControllerAvailableAndShadowLocalesTest.php
Comment (alexander-schranz) [inline Tests/Functional/Integration/ExampleControllerAvailableAndShadowLocalesTest.php:1]: As it is a very individual test case with different tests depending on each other I thought it most make sense to have it in an own class

File: Content/Application/ContentCopier/ContentCopierInterface.php [hunk 8]
```diff
@@ -26,14 +26,18 @@ interface ContentCopierInterface
...
      * @return T
      */
     public function copy(
         ContentRichEntityInterface $sourceContentRichEntity,
         array $sourceDimensionAttributes,
         ContentRichEntityInterface $targetContentRichEntity,
-        array $targetDimensionAttributes
+        array $targetDimensionAttributes,
+        array $data = [],
+        array $ignoredAttributes = []
```
Comment (alexander-schranz) [inline Content/Application/ContentCopier/ContentCopierInterface.php]: While looking now again on this maybe put this together in an `array $options` so this interface don't is so messed up and more future proof. What do you think @Prokyonn @wachterjohannes 

```php
$contentCopier->copy(
    $contentRichEntity,
    $sourceDimensionAttributes,
    $contentRichEntity,
    $targetDimensionAttributes,
    options: [
        'data': [
            'shadowOn' => true,
            'shadowLocale' => 'de',
            'url' => '/test-de',
        ],
        'ignoredAttributes': [
            // ...
        ],
    ] 
)
```

Comment (wachterjohannes) [reply Content/Application/ContentCopier/ContentCopierInterface.php]: sounds okfor me 👍

**Review by wachterjohannes — APPROVED** [review]
For me the code Looks good 👍  if you want to refactor the additional arguments with options - do it! else it would be also OK for me like it is now

File: Content/Application/ContentCopier/ContentCopierInterface.php (hunk 8 above)
Comment (Prokyonn) [reply Content/Application/ContentCopier/ContentCopierInterface.php]: I would also prefer the `options` parameter, the interface would be cleaner and with correct php doctypes the developer gets the same information

**Review by Prokyonn — APPROVED** [review]

File: Content/Application/ContentCopier/ContentCopierInterface.php (hunk 3 above)
Comment (alexander-schranz) [reply Content/Application/ContentCopier/ContentCopierInterface.php]: moved this to a new $options parameter as written above do you want to recheck the comment there?

File: composer.json [hunk 9]
```diff
@@ -65,6 +65,7 @@
         "thecodingmachine/phpstan-strict-rules": "^1.0"
     },
     "conflict": {
+        "coduo/php-matcher: "6.0.12",
```
Comment (alexander-schranz) [inline composer.json:68]: https://github.com/coduo/php-matcher/issues/463

File: composer.json [hunk 10]
```diff
@@ -65,6 +65,7 @@
         "thecodingmachine/phpstan-strict-rules": "^1.0"
     },
     "conflict": {
+        "coduo/php-matcher": "6.0.12",
```
Comment (alexander-schranz) [inline composer.json:68]: conflict because of: https://github.com/coduo/php-matcher/issues/463

**Pull request merged**